spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraSourceRelation.scala

```diff
@@ -0,0 +1,426 @@
+package org.apache.spark.sql.cassandra
+
+import scala.reflect.ClassTag
+
+import org.apache.spark.Logging
+
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.cassandra.CassandraSQLRow.CassandraSQLRowReader
+import org.apache.spark.sql.catalyst.expressions.{Literal, AttributeReference, Expression, Attribute}
+import org.apache.spark.sql.sources._
+import org.apache.spark.sql.types._
+import org.apache.spark.sql.{DataFrame, Row, SQLContext}
+
+import com.datastax.spark.connector._
+import com.datastax.spark.connector.cql.{CassandraConnector, ColumnDef}
+import com.datastax.spark.connector.rdd.{ReadConf, ValidRDDType}
+import com.datastax.spark.connector.writer.{WriteConf, SqlRowWriter}
+
+/**
+ * Table scan type. It allows to create a Relation for given table, keyspace, cluster and user
+ * defined schema. It also allow to pass [[CassandraConnector]], [[ReadConf]] and [[WriteConf]]
+ * explicitly
+ */
+sealed trait ScanType {
```

Comment:
At least, PrunedFilteredScan, PrunedFilterScan, CatalystScan should be implemented. We can move scanType parameter to implicit parameter, so that user can only change it explicitly. e.g. PrunedFilteredScan has some performance issue/error out. It can switch to other scan type.
e.g. PrunedScan does all the filtering internally by Spark SQL, there's less chance to error out. Disadvantage is no filter pushdown.

PrunedFilteredScan leaves to integration to do the filtering( pushdown could be disable, by default it's enabled) This is the default scan type.

CatalystScan filtering using Catalyst internal function to get performance gain, and it hooks into Catalyst query planner directly which potentially gains some performance. As for binary compatibility, it's up to user to decide whether they care about it or not. If it's useless why it's in the API anyway?